### PR TITLE
Symmetric linear multistep integrator

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: integrators/symmetric_linear_multistep_integrator_body.hpp [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+namespace {
+
+// For the Adams-Moulton integrators, see
+// http://www.scholarpedia.org/article/Linear_multistep_method#Methods_for_Hamiltonian_systems  // NOLINT
+// and
+// https://en.wikipedia.org/wiki/Linear_multistep_method#Adams.E2.80.93Moulton_methods.  // NOLINT
+// The latter has a general formula for the coefficients and was used to
+// compute the high-order integrators.
+
+AdamsMoulton<1> const& AdamsMoultonOrder1() {
```
Comment (eggrobin) [inline integrators/symmetric_linear_multistep_integrator_body.hpp]: There is a unique Adams-Moulton method of order `k`. Make this a function template and write specializations. Moreover, I'd like `AdamsMoulton` to be in a separate file: even though we do not implement the actual Adams-Moulton method, it is a different thing which we happen to reuse for the symmetric linear multistep integrator.

File: integrators/symmetric_linear_multistep_integrator_body.hpp [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    t.Increment(h);
+    previous_steps.pop_front();
+    previous_steps.emplace_back();
+    Step& current_step = previous_steps.back();
+    current_step.time = t;
+    current_step.accelerations.resize(dimension);
+
+    // Fill the new step.  We skip the division by ɑk as it is equal to 1.0.
+    double const ɑk = ɑ_[0];
+    CHECK_EQ(ɑk, 1.0);
```
Comment (eggrobin) [inline integrators/symmetric_linear_multistep_integrator_body.hpp]: Do we have to do this in a loop? it seems like it should be done at construction.

File: integrators/symmetric_linear_multistep_integrator_body.hpp [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    typename ODE::SystemState& system_state = down_cast_instance->current_state;
+    std::vector<Position> positions;
+    positions.reserve(dimension);
```
Comment (eggrobin) [inline integrators/symmetric_linear_multistep_integrator_body.hpp]: Allocate outside the loop?

File: numerics/fixed_arrays.hpp [hunk 4]
```diff
@@ -32,6 +32,8 @@ class FixedVector {
 
   operator std::vector<Scalar>() const;
 
+  static int const size = size_;
```
Comment (eggrobin) [inline numerics/fixed_arrays.hpp]: constexpr

File: integrators/symmetric_linear_multistep_integrator.hpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+// Definition of an Adams-Moulton integrator used for computing velocities.
+template<int order_>
+struct AdamsMoulton {
+  static int const order = order_;
+  FixedVector<double, order> numerators;
+  double denominator;
+};
+
+template <typename Position, int order_>
```
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator.hpp]: Done.

File: integrators/symmetric_linear_multistep_integrator.hpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  static int const half_order_ = order_ / 2 + 1;
+  // The velocity is evaluated for a single step, and for a method of order
+  // n a single step has order n + 1.  This declaration gives us the velocity
+  // with order |order_|.
+  static int const velocity_order_ = order_ - 1;
+ public:
+  using ODE = SpecialSecondOrderDifferentialEquation<Position>;
+
+  SymmetricLinearMultistepIntegrator(
+      serialization::FixedStepSizeIntegrator::Kind const kind,
```
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator.hpp]: Done.

File: integrators/symmetric_linear_multistep_integrator.hpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      FixedStepSizeIntegrator<ODE> const& startup_integrator,
+      AdamsMoulton<velocity_order_> const& velocity_integrator,
+      FixedVector<double, half_order_> const& ɑ,
+      FixedVector<double, half_order_> const& β_numerator,
+      double const β_denominator);
```
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator.hpp]: Done.

File: integrators/symmetric_linear_multistep_integrator.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+  void Solve(Instant const& t_final,
+             not_null<IntegrationInstance*> const instance) const override;
```
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator.hpp]: Done.

File: integrators/symmetric_linear_multistep_integrator.hpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+  not_null<std::unique_ptr<IntegrationInstance>> NewInstance(
+    IntegrationProblem<ODE> const& problem,
+    IntegrationInstance::AppendState<ODE> append_state,
+    Time const& step) const;
+
+  static int const order = order_;
```
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator.hpp]: Done.

File: integrators/symmetric_linear_multistep_integrator_body.hpp (hunk 2 above)
Comment (pleroy) [inline integrators/symmetric_linear_multistep_integrator_body.hpp]: Done.

**Pull request merged**